frontend/workflows/projectSelector/src/selector-reducer.tsx

```diff
@@ -28,22 +28,47 @@ const selectorReducer = (state: State, action: Action): State => {
     // User actions.
 
     case "ADD_PROJECTS": {
+      let newState = { ...state };
+
       // a given custom project may already exist in the group so don't trigger a state update for those duplicates
       const uniqueCustomProjects = action.payload.projects.filter(
-        (project: string) => !(project in state[action.payload.group])
+        (project: string) => !(project in newState[action.payload.group])
       );
       if (uniqueCustomProjects.length === 0) {
-        return state;
+        return newState;
       }
-      return {
-        ...state,
-        [action.payload.group]: {
-          ...state[action.payload.group],
-          ...Object.fromEntries(
-            uniqueCustomProjects.map(v => [v, { checked: true, custom: true }])
-          ),
-        },
+
+      newState[action.payload.group] = {
+        ...state[action.payload.group],
+        ...Object.fromEntries(uniqueCustomProjects.map(v => [v, { checked: true, custom: true }])),
       };
+
+      // if the newly added project is in Group.Projects, check to see if we have
+      // project data for it (b/c we don't need to make an API call) and update the state with
+      // its upstreams/downstreams
+      if (action.payload.group === Group.PROJECTS) {
```

Comment:
try to keep happy path unindented. this is clearer to read.

```suggestion
      if (action.payload.group !== Group.PROJECTS) {
        return newState;
      }
      // continue with the other logic
```